packages/engine/Source/Renderer/Texture.js

```diff
@@ -853,94 +919,33 @@ Texture.prototype.copyFrom = function (options) {
 
   let uploaded = false;
   if (!this._initialized) {
-    let pixels;
     if (
       xOffset === 0 &&
       yOffset === 0 &&
-      width === textureWidth &&
-      height === textureHeight
+      width === this._width &&
+      height === this._height
     ) {
       // initialize the entire texture
-      if (defined(arrayBufferView)) {
-        gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
-        gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, false);
-        if (flipY) {
-          pixels = PixelFormat.flipY(
-            arrayBufferView,
-            pixelFormat,
-            pixelDatatype,
-            textureWidth,
-            textureHeight,
-          );
-        } else {
-          pixels = arrayBufferView;
-        }
+      if (defined(source.arrayBufferView)) {
+        loadBufferSource(this, source);
       } else {
-        // Only valid for DOM-Element uploads
-        gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, preMultiplyAlpha);
-        gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, flipY);
-        pixels = source;
+        loadImageSource(this, source);
       }
       uploaded = true;
     } else {
       gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
       gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, false);
-      // initialize the entire texture to zero
-      pixels = PixelFormat.createTypedArray(
-        pixelFormat,
-        pixelDatatype,
-        textureWidth,
-        textureHeight,
-      );
```

Comment:
After those docs burned me the first time around, I'm now relying on the spec rather than MDN.
The WebGL2 spec only mentions _differences_ from WebGL1. The way `null` is handled is mentioned [here](https://registry.khronos.org/webgl/specs/latest/1.0/index.html#TEXIMAGE2D):

> If pixels is null, a buffer of sufficient size initialized to 0 is passed.